src/translator/response_builder.cpp

```diff
@@ -9,6 +9,78 @@ void ResponseBuilder::buildQualityScores(Histories &histories, Response &respons
   qualityEstimator_.computeQualityScores(histories, response);
 }
 
+void buildTagAlignment(Response &response) {
+  // Step 0: set up look-up tables
+  ABORT_IF(response.source.numSentences() == 0, "No sentence in source");
+  std::vector<size_t> tokenOffsetTable(response.source.numSentences(), 0);
+  size_t numSourceTokens = response.source.numWords(0);
+  for (size_t i = 1; i < response.source.numSentences(); i++) {
+    tokenOffsetTable[i] = tokenOffsetTable[i - 1] + response.source.numWords(i);
+    numSourceTokens += response.source.numWords(i);
+  }
+  ABORT_IF(numSourceTokens == 0, "No token in source");
+
+  std::vector<size_t> targetTokenSidTable;
+  std::vector<size_t> targetTokenTidTable;
+  size_t numTargetTokens = 0;
+  for (size_t i = 0; i < response.target.numSentences(); i++) {
+    for (size_t j = 0; j < response.target.numWords(i); j++) {
+      targetTokenSidTable.push_back(i);
+      targetTokenTidTable.push_back(j);
+    }
+    numTargetTokens += response.target.numWords(i);
+  }
+
+  std::vector<size_t> char2TokenTable(response.source.text.size(), 0);
+  std::vector<bool> char2TokenTableValid(response.source.text.size(), 0);
+  for (size_t sentenceId = 0; sentenceId < response.source.numSentences(); sentenceId++) {
+    // Step 0: create or update char-> token table
+    for (size_t s = 0; s < response.source.numWords(sentenceId); s++) {
+      auto sourceByteRange = response.source.wordAsByteRange(sentenceId, s);
+      for (size_t cur = sourceByteRange.begin; cur < sourceByteRange.end; cur++) {
+        char2TokenTable[cur] = s + tokenOffsetTable[sentenceId];
+        char2TokenTableValid[cur] = true;
+      }
+    }
+  }
+
+  for (size_t i = 0; i < char2TokenTable.size(); i++) {
```

Comment:
`char2TokenTable[response.source.text.size()]` is required, as the closing byterange could be the final one. Not having legal value at this index or the size will cause segfault.